### PR TITLE
Generator for sklearn

### PR DESCRIPTION
**Pull request by DrAnaximandre**
**Abstract:**
This PR intends to extend the support of the fuel generators to the sklearn backend.
Although the tests are passing in local (local and async), the async cannot be tested yet on Travis and the test are thus commented on the original PR.
As stated in #40 there was a need for many checks depending the size of the generator of data / data_val, thus leading to a possible code replication.

**Major modifications:**

in `sklearn_backend.py`:

- the `save_params` function has been modified so that a specific format of parameters (list of np.array; specifically encountered in `sklearn.discriminant_analysis.QuadraticDiscriminantAnalysis` types of models) can be inserted in a h5 file.
- the `load_params` function has been modified accordingly to catch and reform the original list of array. The code quality could be improved.
- the `train` function has been modified to deal with fuel generators. The modification are based upon the  `train` function of `keras_backend.py` but differ (no need to call `cm.transform_gen` for instance). 
The 5 cases defined are dealt with (no generator/ generator only for data/ generators for data and data_val: many to many; one to many; many to one). The code seem duplicated and could be factorized in a next part. 

in `test_sklearn_backend.py`:

- a few helper functions ( `close_gen`, `generate_data`,`dump_data`, `make_gen`) have been introduced, based upon their equivalent in  `test_keras_backend.py`. However they do not behave the same due to the difference in classif/regression among others.
- the tests with generator (`test_experiment_fit_gen_nogenval`, `test_experiment_fit_gen_withgenval`) cover the 4 new cases with generator and check that the number of values in the metric (and validation metrics) is as expected. Some more severe tests could be included checking the exactitude of these metrics.

**Minor modifications:**

- for some reason, the data_val was stored in the json in `sklearn_backend.py / fit`. 
- the doc regarding sklearn backend has been slightly modified to fix a typo in #29.

---
This change is [Reviewable](https://reviewable.io/reviews/tboquet/python-alp/44)

### CODE REVIEW

**Comment by tboquet** [comment]
Reviewed 6 of 6 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

_[src/alp/backend/sklearn_backend.py, line 63 at r1](https://reviewable.io:443/reviews/tboquet/python-alp/44#-KVkpSfuEg6Z0T2iWR8I:-KVkpSfuEg6Z0T2iWR8J:ba3dfmt) ([raw file](https://github.com/tboquet/python-alp/blob/7ee788b9e61e05d4fcdfc5f17cde8fe24920ca0d/src/alp/backend/sklearn_backend.py#L63)):_

> ``` Python
>     return SK
> 
> # def recflattenlist(mylist):
> ```

Remove.

---

_[src/alp/backend/sklearn_backend.py, line 79 at r1](https://reviewable.io:443/reviews/tboquet/python-alp/44#-KVkpXv0Im5qGoqpY6EA:-KVkpXv13KzWZe111Nud:ba3dfmt) ([raw file](https://github.com/tboquet/python-alp/blob/7ee788b9e61e05d4fcdfc5f17cde8fe24920ca0d/src/alp/backend/sklearn_backend.py#L79)):_

> ``` Python
>     for k, v in attr.items():
>         if k[-1:] == '_':
>             dict_params[k] = v  # typeconversion(v)
> ```

Remove.

---

_[tests/backend/test_sklearn_backend.py, line 259 at r1](https://reviewable.io:443/reviews/tboquet/python-alp/44#-KVkqGsP640KGh9Cbc6_:-KVkqGsQ7Ukpn12b5c36:b-qmycng) ([raw file](https://github.com/tboquet/python-alp/blob/7ee788b9e61e05d4fcdfc5f17cde8fe24920ca0d/tests/backend/test_sklearn_backend.py#L259)):_

> ``` Python
>         print(self)
> 
>     # def test_experiment_fit_gen_async_nogenval(self, get_model_data_expe):
> ```

Put a TODO for a new travis_container.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/tboquet/python-alp/44)_

**Comment by tboquet** [comment]
Reviewed 1 of 6 files at r1.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

_[src/alp/backend/sklearn_backend.py, line 139 at r1](https://reviewable.io:443/reviews/tboquet/python-alp/44#-KVktgyqIdNtPoOLXvGN:-KVktgyqIdNtPoOLXvGO:bi3gbgv) ([raw file](https://github.com/tboquet/python-alp/blob/7ee788b9e61e05d4fcdfc5f17cde8fe24920ca0d/src/alp/backend/sklearn_backend.py#L139)):_

> ``` Python
>                     out = v[()]
>             setattr(model, k, out)
>         else:
> ```

pragma: no cover.

---

_[src/alp/backend/sklearn_backend.py, line 160 at r1](https://reviewable.io:443/reviews/tboquet/python-alp/44#-KVkt_ZygezI9a_7mSan:-KVkt_ZygezI9a_7mSao:b1c2zib) ([raw file](https://github.com/tboquet/python-alp/blob/7ee788b9e61e05d4fcdfc5f17cde8fe24920ca0d/src/alp/backend/sklearn_backend.py#L160)):_

> ``` Python
>     """
> 
>     if isinstance(v, np.ndarray):  # pragma no cover
> ```

"pragma: no cover".

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/tboquet/python-alp/44)_

**Comment by DrAnaximandre** [comment]
Review status: 4 of 6 files reviewed at latest revision, 5 unresolved discussions.

---

_[src/alp/backend/sklearn_backend.py, line 79 at r1](https://reviewable.io:443/reviews/tboquet/python-alp/44#-KVkpXv0Im5qGoqpY6EA:-KVkvfpFaGOerNwcxvRN:b-7rl0kd) ([raw file](https://github.com/tboquet/python-alp/blob/7ee788b9e61e05d4fcdfc5f17cde8fe24920ca0d/src/alp/backend/sklearn_backend.py#L79)):_
<details><summary>

<i>Previously, tboquet (Thomas Boquet) wrote…</i></summary>



> Remove.
> </details>
> done

---

_[src/alp/backend/sklearn_backend.py, line 139 at r1](https://reviewable.io:443/reviews/tboquet/python-alp/44#-KVktgyqIdNtPoOLXvGN:-KVkvl5dc64s8hSBUs_l:b-7z38px) ([raw file](https://github.com/tboquet/python-alp/blob/7ee788b9e61e05d4fcdfc5f17cde8fe24920ca0d/src/alp/backend/sklearn_backend.py#L139)):_
<details><summary>

<i>Previously, tboquet (Thomas Boquet) wrote…</i></summary>



> pragma: no cover.
> </details>
> removed else pass.

---

_[src/alp/backend/sklearn_backend.py, line 160 at r1](https://reviewable.io:443/reviews/tboquet/python-alp/44#-KVkt_ZygezI9a_7mSan:-KVkvuD_oIXn2SfieDL5:bb7f2mf) ([raw file](https://github.com/tboquet/python-alp/blob/7ee788b9e61e05d4fcdfc5f17cde8fe24920ca0d/src/alp/backend/sklearn_backend.py#L160)):_
<details><summary>

<i>Previously, tboquet (Thomas Boquet) wrote…</i></summary>



> "pragma: no cover".
> </details>
> done

---

_[tests/backend/test_sklearn_backend.py, line 259 at r1](https://reviewable.io:443/reviews/tboquet/python-alp/44#-KVkqGsP640KGh9Cbc6_:-KVkvwUCa5eL9EK1xxHz:bb7f2mf) ([raw file](https://github.com/tboquet/python-alp/blob/7ee788b9e61e05d4fcdfc5f17cde8fe24920ca0d/tests/backend/test_sklearn_backend.py#L259)):_
<details><summary>

<i>Previously, tboquet (Thomas Boquet) wrote…</i></summary>



> Put a TODO for a new travis_container.
> </details>
> done

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/tboquet/python-alp/44)_

**Comment by tboquet** [comment]
Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/tboquet/python-alp/44)_

**Comment by DrAnaximandre** [comment]
Review status: 5 of 6 files reviewed at latest revision, 1 unresolved discussion.

---

_[src/alp/backend/sklearn_backend.py, line 63 at r1](https://reviewable.io:443/reviews/tboquet/python-alp/44#-KVkpSfuEg6Z0T2iWR8I:-KVl4eMr8QtHe9zuKdjk:bb7f2mf) ([raw file](https://github.com/tboquet/python-alp/blob/7ee788b9e61e05d4fcdfc5f17cde8fe24920ca0d/src/alp/backend/sklearn_backend.py#L63)):_
<details><summary>

<i>Previously, tboquet (Thomas Boquet) wrote…</i></summary>



> Remove.
> </details>
> done

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/tboquet/python-alp/44)_

**Comment by tboquet** [comment]
:lgtm: :tada: :rocket: 

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

_Comments from [Reviewable](https://reviewable.io:443/reviews/tboquet/python-alp/44#-:-KVl5alwW-lq6Ve8fcGH:b3t28yh)_
